src/norm/pool.nim

```diff
@@ -0,0 +1,69 @@
+import std/locks
+
+import sqlite
+
+
+type 
+  Pool* = ref object 
+    defaultSize: Natural
+    conns: seq[DbConn]
+    poolExhaustedPolicy: PoolExhaustedPolicy
+    lock: Lock
+  PoolExhaustedError* = object of CatchableError
+  PoolExhaustedPolicy* = enum
+    pepRaise
+    pepExtend
+
+
+func newPool*(defaultSize: Positive, poolExhaustedPolicy = pepRaise): Pool =
+  result = Pool(defaultSize: defaultSize, conns: newSeq[DbConn](defaultSize), poolExhaustedPolicy: poolExhaustedPolicy)
+
+  initLock(result.lock)
+
+  for conn in result.conns.mitems:
+    conn = getDb()
+
+func defaultSize*(pool: Pool): Natural =
+  pool.defaultSize
+
+func size*(pool: Pool): Natural =
+  len(pool.conns)
+
+func pop*(pool: var Pool): DbConn =
+  withLock(pool.lock):
+    if pool.size > 0:
+      result = pool.conns.pop()
+    else:
+      case pool.poolExhaustedPolicy
+      of pepRaise:
+        raise newException(PoolExhaustedError, "Pool exhausted")
+      of pepExtend:
+        result = getDb()
+
+func add*(pool: var Pool, dbConn: DbConn) =
+  withLock(pool.lock):
+    pool.conns.add(dbConn)
+
+func reset*(pool: var Pool) =
+  withLock(pool.lock):
+    while pool.size > pool.defaultSize:
+      var conn = pool.conns.pop()
+      close conn
+
+func close*(pool: var Pool) =
+  withLock(pool.lock):
+    for conn in pool.conns.mitems:
+      close conn
+
+    pool.conns.setLen(0)
+
+template withDb*(pool: var Pool, body: untyped): untyped =
```

Comment:
I feel tempted to allow the user to define the variable name simply because its "less magic" in a sense.
Would extend the template slightly, but make it more behave how I'd anticipate:
```nim
template withDb*(pool: var Pool, db: untyped, body: untyped) =
  block:
    var db = pool.pop()
    try:
      body
    finally:
      recycleConnection(connection)
  ```

Then again, the current way is more consistent with norms previous API for this.

Reply:
> Then again, the current way is more consistent with norms previous API for this.

Exactly. I've just taken the existing ``withDb`` and added a pool to it. So that it takes its ``db`` variable value from the pool and not from the environment.

I think we can add this other flavor of ``withDb`` as well, it's just a whole different story.